macros/src/lib.rs

```diff
@@ -47,31 +44,64 @@ pub fn entry(
         .into();
     }
 
+    let (statics, stmts) = match extract_static_muts(f.block.stmts) {
+        Err(e) => return e.to_compile_error().into(),
+        Ok(x) => x,
+    };
+
     // Rename the function so it is not callable
-    let ident = syn::Ident::new(
-        &format!("_avr_device_rt_{}", f.sig.ident),
+    f.sig.ident = syn::Ident::new(
+        &format!("__avr_device_rt_{}", f.sig.ident),
+        proc_macro2::Span::call_site(),
+    );
+    f.sig.inputs.extend(statics.iter().map(|statik| {
+        let ident = &statik.ident;
+        let ty = &statik.ty;
+        let attrs = &statik.attrs;
+
+        // Note that we use an explicit `'static` lifetime for the entry point arguments. This makes
+        // it more flexible, and is sound here, since the entry will not be called again, ever.
+        syn::parse::<syn::FnArg>(
+            quote::quote!(#[allow(non_snake_case)] #(#attrs)* #ident: &'static mut #ty).into(),
+        )
+        .unwrap()
+    }));
+    f.block.stmts = stmts;
+
+    let tramp_ident = syn::Ident::new(
+        &format!("{}_trampoline", f.sig.ident),
         proc_macro2::Span::call_site(),
     );
+    let ident = &f.sig.ident;
 
-    let attrs = f.attrs;
-    let block = f.block;
-    let stmts = block.stmts;
-    let unsafety = f.sig.unsafety;
+    let resource_args = statics
+        .iter()
+        .map(|statik| {
+            let (ref cfgs, ref attrs) = extract_cfgs(statik.attrs.clone());
+            let ident = &statik.ident;
+            let ty = &statik.ty;
+            let expr = &statik.expr;
+            quote::quote! {
+                #(#cfgs)*
+                {
+                    #(#attrs)*
+                    static mut #ident: #ty = #expr;
+                    &mut #ident
+                }
+            }
+        })
+        .collect::<Vec<_>>();
 
     quote::quote! (
-        #[cfg(not(any(doc, target_arch = "avr")))]
-        compile_error!(
-            "Ensure that you are using an AVR target! You may need to change \
-       directories or pass a --target flag to cargo. See
-       https://github.com/Rahix/avr-device/pull/41 for more details."
-        );
-
-        #(#attrs)*
         #[doc(hidden)]
         #[export_name = "main"]
-        pub #unsafety extern "C" fn #ident() -> ! {
-            #(#stmts)*
+        pub unsafe extern "C" fn #tramp_ident() {
+            #ident(
+                #(#resource_args),*
+            )
         }
+
+        #f
```

Comment:
Previously we had the `#[doc(hidden)]` attribute on the main function to hide it under all circumstances.  I see you've left it on the trampoline, but shouldn't we also add it to the function itself?  Especially because we change its name to `__avr_device_rt_main` I don't see any reason not to...